### PR TITLE
Trait impls for tuples of arity <= 12

### LINKED ISSUE #398: add implementation of 5-tuple for tuple_combinations?
I was using tuple_combinations for iterating over various card combinations and ran into the 4-tuple limitation when trying to iterate over all possible 5-card combinations of a deck of 52 cards.  Any chance you'd add a five-tuple implementation or should I use another method?

### LINKED ISSUE #428: HomogeneousTuple / TupleCollect for N > 4 ?
It would be useful to have HomogeneousTuple / TupleCollect for N larger than 4.
E.g. I often use it with `splitn` to avoid allocations, like:
```rust
s.splitn(N, pattern).collect_tuple()
```
instead of 
```rust
if let [/* ... */] = pattern.splitn(N, pattern).collect_vec().as_slice() {
	Some(/* ... */)
} else {
	None
}
```
And often my N is higher than 4.

### PR DESCRIPTION
**Pull request by jswrenn**
I stopped at 12, because that's where libcore stops.

fixes #428
fixes #398

### CODE REVIEW

File: src/tuple_impl.rs [hunk 1]
```diff
@@ -280,23 +280,14 @@ macro_rules! impl_tuple_collect {
...
-                            break;
-                        };
-                    )*
-                    return Some(($($Y),*,))
-                }
 
-                return None;
+                Some(($(
+                    { let $Y = iter.next()?; $Y },
+                )*))
```
Comment (jswrenn) [inline src/tuple_impl.rs:290]: I *think* this modification is semantically equivalent, but I'd love if someone else could check it over before I merge.

**Comment by jswrenn** [comment]
Seems like a random CI failure?

bors try

**Comment by phimuemue** [comment]
bors try

File: src/tuple_impl.rs (hunk 1 above)
Comment (phimuemue) [reply src/tuple_impl.rs:290]: Looks good to me. Godbolt even seems to generate very similar code in `--release`.

off-topic: I do not have much time to spare at the moment, so I won't be very responsive.

**Pull request merged**